lsp/nls/src/requests/completion.rs

```diff
@@ -29,18 +29,109 @@ use crate::{
 // and stop when there is nothing else on the path
 
+struct IdentWithMeta {
+    ident: Ident,
+    ty: Types,
+    item: Option<LinearizationItem<Types>>,
+}
+
+impl From<Ident> for IdentWithMeta {
+    fn from(ident: Ident) -> Self {
+        IdentWithMeta {
+            ident,
+            ty: Types(TypeF::Dyn),
+            item: None,
+        }
+    }
+}
+
+impl From<&str> for IdentWithMeta {
+    fn from(ident: &str) -> Self {
+        IdentWithMeta {
+            ident: Ident::from(ident),
+            ty: Types(TypeF::Dyn),
+            item: None,
+        }
+    }
+}
+
+impl IdentWithMeta {
+    fn compute_detail(&self) -> String {
+        self.item
+            .as_ref()
+            .and_then(|item| {
+                item.meta.as_ref().and_then(|meta| {
+                    meta.types
+                        .as_ref()
+                        .map(|ty| ty.types.to_string())
+                        .or_else(|| meta.contracts_to_string())
+                })
+            })
+            .unwrap_or_else(|| self.ty.to_string())
+    }
+
+    fn compute_completion_item_kind(&self) -> CompletionItemKind {
+        match &self.ty {
+            ty if ty.is_function_type() => CompletionItemKind::Function,
+            _ => CompletionItemKind::Property,
+        }
+    }
+
+    fn to_lsp_completion_item(&self) -> CompletionItem {
+        /// Attach quotes to a non-ASCII string
+        fn adjust_name(name: &str) -> String {
+            if name.is_ascii() {
+                String::from(name)
+            } else {
+                format!("\"{}\"", name)
+            }
+        }
+        let doc = || {
+            let item = self.item.as_ref()?;
+            let meta = item.meta.as_ref()?;
+            let doc = meta.doc.as_ref()?;
+            let doc =
+                Documentation::String(format!("{}\nMerge Priority {}", doc.clone(), meta.priority));
+            Some(doc)
+        };
+        CompletionItem {
+            label: adjust_name(self.ident.label()),
+            detail: Some(self.compute_detail()),
+            kind: Some(self.compute_completion_item_kind()),
+            documentation: doc(),
+            ..Default::default()
+        }
+    }
+}
+
 /// Find the record field associated with a particular ID in the linearization
 /// using lexical scoping rules.
 fn find_fields_from_term_kind(
     linearization: &Completed,
     id: ItemId,
     path: &mut Vec<Ident>,
-) -> Option<Vec<Ident>> {
+) -> Option<Vec<IdentWithMeta>> {
     let item = linearization.get_item(id)?;
     match item.kind {
         TermKind::Record(ref fields) => {
             if path.is_empty() {
-                Some(fields.keys().cloned().collect())
+                Some(
+                    fields
+                        .keys()
+                        .cloned()
+                        .map(|ident| {
+                            // These unwraps are safe
```

Comment:
The important information to add here is _why_ they are safe :slightly_smiling_face: . Because we ensured somewhere else in the code that all those indices are indeed valid, I guess. But if I refactor the code I have no idea what I have to do to preserve this property.